src/server/editorServices.ts

```diff
@@ -0,0 +1,1646 @@
+                for (var i = 0, len = referencingProjects.length; i < len; i++) {
+                    referencingProjects[i].removeReferencedFile(info);
+                }
+            }
+            this.printProjects();
+        }
+
+        addOpenFile(info: ScriptInfo) {
+            this.findReferencingProjects(info);
+            if (info.defaultProject) {
+                this.openFilesReferenced.push(info);
+            }
+            else {
+                // create new inferred project p with the newly opened file as root
+                info.defaultProject = this.createInferredProject(info);
+                var openFileRoots: ScriptInfo[] = [];
+                // for each inferred project root r
+                for (var i = 0, len = this.openFileRoots.length; i < len; i++) {
+                    var r = this.openFileRoots[i];
+                    // if r referenced by the new project
+                    if (info.defaultProject.getSourceFile(r)) {
+                        // remove project rooted at r
+                        this.inferredProjects =
+                        copyListRemovingItem(r.defaultProject, this.inferredProjects);
+                        // put r in referenced open file list
+                        this.openFilesReferenced.push(r);
+                        // set default project of r to the new project 
+                        r.defaultProject = info.defaultProject;
+                    }
+                    else {
+                        // otherwise, keep r as root of inferred project
+                        openFileRoots.push(r);
+                    }
+                }
+                this.openFileRoots = openFileRoots;
+                this.openFileRoots.push(info);
+            }
+        }
+
+        closeOpenFile(info: ScriptInfo) {
+            var openFileRoots: ScriptInfo[] = [];
+            var removedProject: Project;
+            for (var i = 0, len = this.openFileRoots.length; i < len; i++) {
+                // if closed file is root of project
+                if (info == this.openFileRoots[i]) {
+                    // remove that project and remember it
+                    removedProject = info.defaultProject;
+                }
+                else {
+                    openFileRoots.push(this.openFileRoots[i]);
+                }
+            }
+            this.openFileRoots = openFileRoots;
+            if (removedProject) {
+                // remove project from inferred projects list
+                this.inferredProjects = copyListRemovingItem(removedProject, this.inferredProjects);
+                var openFilesReferenced: ScriptInfo[] = [];
+                var orphanFiles: ScriptInfo[] = [];
+                // for all open, referenced files f
+                for (var i = 0, len = this.openFilesReferenced.length; i < len; i++) {
+                    var f = this.openFilesReferenced[i];
+                    // if f was referenced by the removed project, remember it

```

Comment:
These "remember it" comments, I'd prefer if you said something above the logic like "Find all files that were referenced from the removed project so that we can reinterpret them as newly opened."